_posts/2024-04-26-modules-101.markdown

```diff
@@ -0,0 +1,137 @@
+---
+layout: post
+title:  "Valkey modules 101"
+authors: 
+  - dmitrypol
+date: 2024-04-26 01:01:01 -0700
+categories: update
+---
+
+## What are Valkey (or Redis) modules?  
+
+Modules were first announced at RedisConf 2016 and officially introduced with Redis 4.0.  The idea of modules is to keep the core functionality lean and add extra features (such as new commands and data types) as needed.  Over the years there were a number of modules developed for Redis.  Valkey project plans to build on this foundation.  
+
+Modules can be written in C or other languages that have C bindings.  In this article we will go through the process of building simple modules in C and Rust (using the Valkey Module Rust SDK).  This articles expects the audience to be at least somewhat familiar with C, Rust and Redis/Valkey.  
+
+## Hello World module in C
+
+If we clone [Valkey](https://github.com/valkey-io/valkey) repo we will find numerous examples in `src/modules`.  Let's create a new file `module1.c` in the same folder.  
+
+```c
+#include "../valkeymodule.h"
+
+int hello(ValkeyModuleCtx *ctx, ValkeyModuleString **argv, int argc) {
+    VALKEYMODULE_NOT_USED(argv);
+    VALKEYMODULE_NOT_USED(argc);
+    return ValkeyModule_ReplyWithSimpleString(ctx, "world1");
+}
+
+int ValkeyModule_OnLoad(ValkeyModuleCtx *ctx, ValkeyModuleString **argv, int argc) {
+    VALKEYMODULE_NOT_USED(argv);
+    VALKEYMODULE_NOT_USED(argc);
+    if (ValkeyModule_Init(ctx,"module1",1,VALKEYMODULE_APIVER_1) 
+        == VALKEYMODULE_ERR) return VALKEYMODULE_ERR;
+    if (ValkeyModule_CreateCommand(ctx,"module1.hello", hello,"",0,0,0) 
+        == VALKEYMODULE_ERR) return VALKEYMODULE_ERR;
+    return VALKEYMODULE_OK;
+}
+```
+
+All we are doing is initalizing `module1` and creating a command `hello` that returns `world1`.  Now we need to update `src/modules/Makefile`
+
+```makefile
+all: ... module1.so
+
+module1.xo: ../valkeymodule.h
+
+module1.so: module1.xo
+	$(LD) -o $@ $^ $(SHOBJ_LDFLAGS) $(LIBS) -lc
+```
+
+Run `make` inside `src/modules` folder.  We will see numerous *.so and *.xo files created in `src/modules` including `module1.so`.  
+
+
+## Hello World module in Rust
+
+If we clone the [valkeymodule-rs](https://github.com/valkey-io/valkeymodule-rs) repo we will find `examples` folder with numerous .rs files.  Valkeymodule-rs is a recent fork of redismiodule-rs and there are still references to Redis in it.  We plan to update this repo and add it to [crates.io](https://crates.io/) but for now we will use it as is.  
```

Comment:
Ideally we would use the valkeymodule-rs SDK correctly in this blog post, which is to take a dependency on it and then build it. We don't really want people using the sdk this way, and in the long term people might get confused by having this legacy post.

Reply:
Can we also cleanup the references to redis before publishing this? I think it would be better for branding to limit the references to Redis in this post.

Reply:
Agree on branding.  I am working to get SDK into crates.io but I want to cleanup before publishing it as crate.

Reply:
@madolson - per our chat I published https://crates.io/crates/valkey-module and updated this blog.